README.md
# Pallet AWS Provider

[![Build Status](https://travis-ci.org/pallet/pallet-aws.svg?branch=0.2.7)](https://travis-ci.org/pallet/pallet-aws)

A [pallet](http://palletops.com/) provider for [AWS](http://aws.amazon.com/), based on
the Amazon java SDK.

The EC2 provider uses the `:pallet-ec2` key.

## Usage

Add the following to your dependencies:

```clj
[com.palletops/pallet-aws "0.2.7"]
[org.slf4j/jcl-over-slf4j "1.7.5"]
```

Warning: versions 0.2.2 and prior contain a nasty bug that will remove
all the nodes in your account if you converge any group down to 0 nodes.

The underlying AWS SDK uses `commons-logging` for logging.  This means
you will need to add a dependency that provides the commons logging
interface.  Use `[commons-logging "1.1.3"]` if you want to use commons
logging, or `[org.slf4j/jcl-over-slf4j "1.7.5"]` if you are using slf4
(eg. with logback).

## Node-Spec Options

The `node-spec` can be used to control EC2 specific features:

### Image

The ami to use for the image is passed as a string to the `:image-id`
key in the node-spec's `:image`.

```clj
(node-spec ... :image {:image-id "ami-35792c5c"
                                 :os-family :amzn-linux
                                 :os-version "2013.09"
                                 :login-user "ec2-user"})
```

You must specify the `:os-family`, `:os-version` and `:login-user`
with ami specific information.

### Availability Zones

The availability zone can be specifed in the `:location` key of the node-spec.

```clj
(node-spec ... :location {:location-id "us-east-1d"})
```

### KeyPairs

By default, pallet creates a keypair based on the admin username and key credentials.

To reuse an existing EC2 keypair, pass the name of the keypair to the
`:key-name` key of the `:provider` specific map.

```clj
(node-spec ... :provider {:pallet-ec2 {:key-name "mykey"}})
```

This keypair is authorised on the node when it is created, and is used
by pallet to run the `:bootstrap` phase.

### Block devices and EBS

The provider specific `block-device-mappings` key can be used to set up block devices.
It takes a sequence of maps with the following keys:

`:device-name`
: The device name exposed to the instance (for example, "/dev/sdh").

`:virtual-name`
: The virtual device name (ephemeral[0..3]). The number of available
instance store volumes depends on the instance type. After you connect
to the instance, you must mount the volume.

`:ebs :snapshot-id`
: The ID of the snapshot.

`:ebs :volume-size`
: The size of the volume, in gigabytes.

`:ebs :delete-on-termination`
: Indicates whether to delete the volume on instance termination.

`:ebs :volume-type`
: The volume type ("standard", "gp2" for SSD, or "io1" for SSD with provisioned IOPS)

`:ebs :iops`
: For "io1" volumes, the number of I/O operations per second (IOPS)
that the volume supports. Range is 100 to 4000.

`:no-device`
: Suppresses the device mapping.

### Network Interfaces

The provider specific `:network-interface` key takes a sequence of
maps with the following keys (and key paths):

`:network-interface-id`
: An existing interface to attach to a single instance. Requires n=1
instances.

`:device-index`
: The device index. Applies both to attaching an existing network
interface and creating a network interface.  If you are specifying a
network interface in the request, you must provide the device index.

`:subnet-id`
: The subnet ID. Applies only when creating a network interface.

`:description`
: A description. Applies only when creating a network interface.

`:private-ip-address`
: The primary private IP address. Applies only when creating a network
interface. Requires n=1 network interfaces in launch.

`:security-group-id`
: The ID of the security group. Applies only when creating a network interface.

`:delete-on-termination`
: Indicates whether to delete the network interface on instance termination.

`:private-ip-addresses [] :private-ip-address`
: The private IP address. `:private-ip-addresses` takes a sequence of
maps to specify explicit private IP addresses for a network interface,
but only one private IP address can be designated as primary.  Only
one private IP address can be designated as primary. Therefore, you
can't specify this parameter if `:private-ip-addresses :primary` is set
to true and `:private-ip-address` is set to an IP address.

`:private-ip-addresses [] :primary`
: Indicates whether the private IP address is the primary private IP address.

`:secondary-private-ip-addresscount`
: The number of private IP addresses to assign to the network
interface. For a single network interface, you can't specify this
option and specify more than one private IP address using
`:private-ip-address`.

`:associate-public-ip-address`
: Indicates whether to assign a public IP address to an instance in a
VPC. The public IP address is associated with a specific network
interface. If set to true, the following rules apply: Can only be
associated with a single network interface with the device index
of 0. You can't associate a public IP address with a second network
interface, and you can't associate a public IP address if you are
launching more than one network interface.  Can only be associated
with a new network interface, not an existing one. Default: If
launching into a default subnet, the default value is true. If
launching into a nondefault subnet, the default value is false.

### IAM Roles

An IAM profile can be specified on the `:iam-instance-profile`
provider specific key. It is a map with one of the following keys:

`:arn`
: The Amazon Resource Name (ARN) of the IAM instance profile to
associate with the instances.

`:name`
: The name of the IAM Instance Profile (IIP) to associate with the
instances.

An example node-spec might look like:

``` clj
;; Remember to use the "Instance Profile ARN"
(def ^:private instance-arn
  "arn:aws:iam::000000000000:instance-profile/SomeName")

(def iam-node
  (node-spec
   :image {:os-family :ubuntu
           :os-version-matches "12.04"
           :os-64-bit true
           :image-id "us-east-1/ami-e2861d8b"}
   :hardware {:min-cores 2 :min-ram 512}
   :network {:inbound-ports [22 1234]}
   :qos {:enable-monitoring true}
   :provider
   {:pallet-ec2 {:iam-instance-profile {:name instance-arn}}}))
```

### Other

The `[:provider :pallet-ec2]` path can be used to specify other EC2
specific functionality.

`:user-data`
: The Base64-encoded MIME user data for the instances.


`:placement :group-name`
: The name of an existing placement group

`:placement :tenancy`
: The tenancy of the instance. An instance with a tenancy of `"dedicated"`
runs on single-tenant hardware and can only be launched into a VPC.

`:kernel-id`
: The ID of the kernel.  It is recommended that you use PV-GRUB
instead of kernels and RAM disks. For more information, see PV-GRUB: A
New Amazon Kernel Image in the Amazon Elastic Compute Cloud User
Guide.

`:ramdisk-id`
: The ID of the RAM disk.

`:monitoring :enabled`
: Enables monitoring for the instance.

`:subnet-id`
: the ID of the subnet to launch the instance into.

`:disable-api-termination`
: If you set this parameter to true, you can't terminate the instance
using the Amazon EC2 console, CLI, or API; otherwise, you can. If you
set this parameter to true and then later want to be able to terminate
the instance, you must first change the value of the
disableApiTermination attribute to false using
ModifyInstanceAttribute. Alternatively, if you set
InstanceInitiatedShutdownBehavior to terminate, you can terminate the
instance by running the shutdown command from the instance.

`:instance-initiated-shutdown-behavior`
: Indicates whether an instance stops or terminates when you initiate
shutdown from the instance (using the operating system command for
system shutdown). Either "stop" or "terminate".

`::private-ip-address`
: The primary IP address. You must specify a value from the IP address
range of the subnet.  Only one private IP address can be designated as
primary. Therefore, you can't specify this parameter if
PrivateIpAddresses.n.Primary is set to true and
PrivateIpAddresses.n.PrivateIpAddress is set to an IP address.

`:client-token`
: Unique, case-sensitive identifier you provide to ensure idempotency
of the request. For more information, see How to Ensure Idempotency in
the Amazon Elastic Compute Cloud User Guide.

`:ebs-optimized`
: Indicates whether the instance is optimized for EBS I/O. This optimization provides dedicated throughput to Amazon EBS and an optimized configuration stack to provide optimal Amazon EBS I/O performance. This optimization isn't available with all instance types. Additional usage charges apply when using an EBS-optimized instance.



## Access to Arbitrary AWS SDK Funtions

The provide provides an `execute` function that can be used to execute
arbitrary functions, described as [`awaze`](https://github.com/pallet/awaze) data maps.


## License

Copyright © 2013 Hugo Duncan

Distributed under the Eclipse Public License.


ReleaseNotes.md
## 0.2.7

- Add t2.large static information

## 0.2.6

- Update to pallet-aws-ops 0.2.2

- Enable use of :network-interfaces
  The subnet ID and groups were not being handled correctly.

- Report private DNS if public DNS not set

- Fix block device doc in README

## 0.2.5

- Fix generation of signed s3 requests

## 0.2.4

- Make provider work on non-default VPC
  Use security group ids rahter than security group names.

- Update to pallet-aws-ops 0.2.1

- Add jcl-over-slf4j to default dependencies in readme

- Add the new generation ec2 instance types

## 0.2.3

- Fix bug in destroy-nodes-in-group
  The bug would remove all noes in an account when any group was converged
  to 0 nodes.

  Closes #7

## 0.2.2

- Fix destroy-node implementation
  Was incorrectly using :instance-id rather than :instance-ids.

- Remove hardware-name
  The name is translated in awaze, so can be passed directly here.

  Closes #1

- Update to pallet-aws-ops 0.2.0

## 0.2.1

- Add blobstore implementation
  Adds the :pallet-s3 provider for blobstore.

  Closes #4

# 0.2.0

- Update to pallet 0.8.0-RC.9

- Tags on new instances reflected in Nodes

- Tag instance names

- Add ami-info function to parse ami descriptions

- Remove AMI description parsing for Node info

- Implement ComputeServiceProperties

# 0.1.1

- Change describe-key-pairs to use :key-names

- Split Execute protocol into separate namespace

- Fix creating of launch keypair

- Fix link to awaze in the readme

# 0.1.0

- Initial version


test/pallet/compute/ec2_test.clj
(ns pallet.compute.ec2-test
  (:require
   [clojure.test :refer :all]
   [pallet.api :refer [group-spec node-spec]]
   [pallet.compute.ec2 :refer :all]))

(deftest launch-options-test
  (is (= {:security-group-ids ["sg"],
          :key-name "kn",
          :max-count 1,
          :min-count 1,
          :image-id "i"}
         (launch-options
          1
          (group-spec :gn :node-spec (node-spec :image {:image-id "i"}))
          "sg" "kn"))
      "minimal test")
  (is (= {:security-group-ids ["sg"],
          :key-name "kn",
          :max-count 1,
          :min-count 1,
          :image-id "i"
          :instance-type "m1.small"}
         (launch-options
          1
          (group-spec :gn
            :node-spec (node-spec :hardware {:hardware-id "m1.small"}
                                  :image {:image-id "i"}))
          "sg" "kn"))
      "hardware id test")
  (is (= {:security-group-ids ["sg"],
          :key-name "kn",
          :max-count 1,
          :min-count 1,
          :image-id "i"
          :placement {:availability-zone "us-east-1d"}}
         (launch-options
          1
          (group-spec :gn
            :node-spec (node-spec
                        :image {:image-id "i"}
                        :location {:location-id "us-east-1d"}))
          "sg" "kn"))
      "availability zone")
  (is (= {:security-group-ids ["sg"],
          :key-name "kn",
          :max-count 1,
          :min-count 1,
          :image-id "i"
          :block-device-mapping [{:device-name "/dev/sdh"}]}
         (launch-options
          1
          (group-spec :gn
            :node-spec
            (node-spec
             :image {:image-id "i"}
             :provider {:pallet-ec2
                        {:block-device-mapping
                         [{:device-name "/dev/sdh"}]}}))
          "sg" "kn"))
      "block-device-mapping")
  (is (= {:key-name "kn",
          :max-count 1,
          :min-count 1,
          :image-id "i"
          :network-interfaces [{:device-index 0
                                :subnet-id "subnet-abcdef77"
                                :groups ["sg-abcdef88"]
                                :associate-public-ip-address true
                                :delete-on-termination true}]}
         (launch-options
          1
          (group-spec :gn
            :node-spec
            (node-spec
             :image {:image-id "i"}
             :provider {:pallet-ec2
                        {:network-interfaces
                         [{:device-index 0
                                :subnet-id "subnet-abcdef77"
                                :groups ["sg-abcdef88"]
                                :associate-public-ip-address true
                                :delete-on-termination true}]}}))
          "sg" "kn"))
      "network-interfaces"))

(deftest instance-tags-test
  (is (= [{:key "pallet-group", :value "abcd"}
          {:key "pallet-image", :value "{:image-id \"ami\"}"}
          {:key "Name", :value "abcd_1-2-3-4"}]
         (instance-tags {:group-name "abcd" :image {:image-id "ami"}}
                        {:public-ip-address "1.2.3.4"})) ))

(deftest ids-tags-test
  (is (= [["id" {:key "Name", :value "abcd_1-2-3-4"}]]
         (id-tags {:instance-id "id" :public-ip-address "1.2.3.4"}
                  [{:key "Name", :value "abcd_1-2-3-4"}])) ))


test/pallet/compute/ec2/ami_test.clj
(ns pallet.compute.ec2.ami-test
  (:require
   [clojure.test :refer :all]
   [pallet.compute.ec2.ami :refer [parse]]))

(deftest parse-test
  (is (= {:os-family :amzn-linux
          :os-version "2013.09.0"
          :user {:username "ec2-user"}}
         (parse {:owner-id "137112412989"
                 :name "amzn-ami-pv-2013.09.0.x86_64-ebs"
                 :description "Amazon Linux AMI x86_64 PV EBS"}))))

(deftest os-parser-test
  (testing "amzn linux"
    (is (= {:os-family :amzn-linux :os-version "0.9.7-beta"
            :user {:username "ec2-user"}}
           (parse {:owner-id "137112412989"
                   :description "Amazon Linux AMI.*"
                   :name "137112412989/amzn-ami-0.9.7-beta.i386-ebs"})))
    (is (= {:os-family :amzn-linux :os-version "0.9.7-beta"
            :user {:username "ec2-user"}}
           (parse {:owner-id "137112412989"
                   :description "Amazon Linux AMI.*"
                   :name "137112412989/amzn-ami-0.9.7-beta.x86_64-ebs"})))
    (is (= {:os-family :amzn-linux :os-version "0.9.7-beta"
            :user {:username "ec2-user"}}
           (parse
            {:owner-id "137112412989"
             :description "Amazon Linux AMI.*"
             :name "amzn-ami-us-east-1/amzn-ami-0.9.7-beta.x86_64.manifest.xml"})))
    (is (= {:os-family :amzn-linux :os-version "0.9.7-beta"
            :user {:username "ec2-user"}}
           (parse
            {:owner-id "137112412989"
             :description "Amazon Linux AMI.*"
             :name "amzn-ami-us-east-1/amzn-ami-0.9.7-beta.i386.manifest.xml"}))))
  (testing "amazon centos"
    (is (= {:os-family :centos :os-version "5.4"}
           (parse {:owner-id "137112412989"
                   :description "Amazon Centos"
                   :name "amazon/EC2 CentOS 5.4 HVM AMI"}))))
  (testing "canonical"
    (is
     (=
      {:os-family :ubuntu :os-version "12.04" :user {:username "ubuntu"}}
      (parse
       {:owner-id "099720109477"
        :name "ubuntu/images-milestone/ubuntu-precise-12.04-beta1-i386-server-20120229.1"}))))
  (testing "rightscale"
    (is (= {:os-family :centos :os-version "5.4"}
           (parse
            {:owner-id "411009282317"
             :name "rightscale-us-east/CentOS_5.4_x64_v4.4.10.manifest.xml"}))))
  (testing "rightimage"
    (is (= {:os-family :ubuntu :os-version "9.10"}
           (parse
            {:owner-id "411009282317"
             :name "411009282317/RightImage_Ubuntu_9.10_x64_v4.5.3_EBS_Alpha"})))
    (is (= {:os-family :windows :os-version "2008"}
           (parse {:owner-id "411009282317"
                   :name "411009282317/RightImage_Windows_2008_x64_v5.5.5"})))))


test/pallet/compute/ec2/support_test.clj
(ns pallet.compute.ec2.support-test
  (:require
   [clojure.core.async :refer [<!! chan]]
   [clojure.test :refer :all]
   [com.palletops.aws.api :as api]
   [com.palletops.aws.vpc :refer :all]
   [pallet.compute :refer [nodes]]
   [pallet.crates.test-nodes :as test-nodes]
   [pallet.test-env
    :refer [*compute-service* *node-spec-meta* test-env]]
   [pallet.test-env.project :as project]))

(test-env test-nodes/node-specs project/project)

(deftest ^:support nodes-test
  (let [{:keys [credentials] :as props} (service-properties *compute-service*)]
    (is (nodes *compute-service*))))


release.sh
#!/bin/bash

# release pallet-aws

if [[ $# -lt 3 ]]; then
  echo "usage: $(basename $0) previous-version new-version next-version" >&2
  exit 1
fi

previous_version=$1
version=$2
next_version=$3

echo ""
echo "Start release of $version, previous version is $previous_version"
echo ""
echo ""

lein do clean, with-profile +no-checkouts test && \
git flow release start $version || exit 1

lein with-profile +release set-version ${version} :previous-version ${previous_version} \
  || { echo "set version failed" >2 ; exit 1; }

echo ""
echo ""
echo "Changes since $previous_version"
git --no-pager log --pretty=changelog $previous_version..
echo ""
echo ""
echo "Now edit project.clj, ReleaseNotes and README"

$EDITOR project.clj
$EDITOR ReleaseNotes.md
$EDITOR README.md

echo -n "commiting project.clj, release notes and readme.  enter to continue:" \
&& read x \
&& git add project.clj ReleaseNotes.md README.md \
&& git commit -m "Updated project.clj, release notes and readme for $version" \
&& echo -n "Peform release.  enter to continue:" && read x \
&& lein do clean, with-profile +no-checkouts test, deploy clojars \
&& rm -f pom.xml \
&& git flow release finish $version \
&& echo "Now push to github. Don't forget the tags!" \
&& lein with-profile +doc doc \
&& lein with-profile +release set-version ${next_version} \
&& git add project.clj \
&& git commit -m "Updated version for next release cycle"


src/pallet/blobstore/s3.clj
(ns pallet.blobstore.s3
  "Amazon s3 provider for pallet"
  (:require
   [clojure.java.io :refer [input-stream]]
   [clojure.string :as string]
   [com.palletops.awaze.s3 :as s3]
   [com.palletops.aws.api :as aws]
   [pallet.blobstore :as blobstor]
   [pallet.blobstore.implementation :as implementation]
   [pallet.compute.jvm :as jvm]))

(defn buckets [api credentials]
  "Return a sequence of maps for containers"
  (aws/execute
     api
     (s3/list-buckets-map
      credentials
      {})))

(defn bucket-for-name
  "Filter buckets for the given bucket name."
  [buckets bucket-name]
  (first (filter #(= bucket-name (:name %)) buckets)))

(defn create-bucket
  "Create a bucket with the given bucket name."
  [api credentials bucket-name]
  (aws/execute
     api
     (s3/create-bucket-map
      credentials
      {:bucket-name bucket-name})))

(defn put-object-request [])
(defmulti put-object-request
  "Return a partial put-object request map for the payload"
  type)

(defmethod put-object-request :default
  [x]
  {:input-stream (input-stream x)})

(defmethod put-object-request java.io.File
  [file]
  {:file file})

(defmethod put-object-request String
  [s]
  {:input-stream (java.io.ByteArrayInputStream.
                  (.getBytes s java.nio.charset.StandardCharsets/UTF_8))})

(defn put-object
  "Put a payload into a s3 container at path."
  [api credentials bucket-name key payload]
  {:pre [payload]}
  (aws/execute
     api
     (s3/put-object-map
      credentials
      (merge {:bucket-name bucket-name
              :key key}
             (put-object-request payload)))))

(defn one-hour-from-now
  []
  (-> (doto (java.util.Calendar/getInstance)
        (.setTime (java.util.Date.))
        (.add java.util.Calendar/HOUR 1))
      .getTime))

(defn generate-presigned-url
  "Return a pre-signed url."
  [api credentials bucket-name key request-map]
  (aws/execute
     api
     (s3/generate-presigned-url-map
      credentials
      {:bucket-name bucket-name
       :key key
       :expiration (one-hour-from-now)
       :method (-> (:method request-map :GET) name string/upper-case)
       :response-headers (dissoc request-map :method)})))


(deftype S3Service [credentials api]
  pallet.blobstore/Blobstore

  (sign-blob-request [blobstore container path request-map]
    (let [request (generate-presigned-url
                   api credentials container path request-map)]
      {:endpoint request}))

  (put-file [blobstore container path file]
    (when-not (bucket-for-name (buckets api credentials) container)
      (create-bucket api credentials container))
    (put-object api credentials container path file))

  (put [blobstore container path payload]
    (when-not (bucket-for-name (buckets api credentials) container)
      (create-bucket api credentials container))
    (put-object api credentials container path payload))

  (containers [blobstore]
    (map :name (buckets api credentials)))

  (close [blobstore]))

(defmethod implementation/service :pallet-s3
  [provider {:keys [identity credential endpoint api]
             :or {endpoint "US_EAST_1"}}]
  (let [credentials {:access-key identity :secret-key credential
                     :endpoint endpoint}
        api (or api (aws/start {}))]
    (S3Service.
     credentials
     api)))


src/pallet/compute/ec2/static.clj
(ns pallet.compute.ec2.static
  "Static data for EC2 (data not available in ec2 apis)")

(defn GiB [n] (* 1024 n))

(def instance-types
  {;;; Burstable Performance Instances

   :t2.micro
   {:ram (GiB 1)
    :cpus [{:cores 1 :speed 1}]
    :disks []
    :32-bit true
    :64-bit true
    :io :low
    :ebs-optimized false}

   :t2.small
   {:ram (GiB 2)
    :cpus [{:cores 1 :speed 1}]
    :disks []
    :32-bit true
    :64-bit true
    :io :low
    :ebs-optimized false}

   :t2.medium
   {:ram (GiB 4)
    :cpus [{:cores 2 :speed 1}]
    :disks []
    :32-bit true
    :64-bit true
    :io :low
    :ebs-optimized false}

   :t2.large
   {:ram (GiB 8)
    :cpus [{:cores 2 :speed 1}]
    :disks []
    :32-bit true
    :64-bit true
    :io :low
    :ebs-optimized false}

   ;;; General Purpose

   :m3.medium
   {:ram (GiB 3.75)
    :cpus [{:cores 1 :speed 3}]
    :disks [{:size 4}]
    :64-bit true
    :io :moderate
    :ebs-optimized false}

   :m3.large
   {:ram (GiB 7.5)
    :cpus [{:cores 2 :speed 3.25}]
    :disks [{:size 32}]
    :64-bit true
    :io :moderate
    :ebs-optimized false}

   :m3.xlarge
   {:ram (GiB 15)
    :cpus [{:cores 4 :speed 3.25}]
    :disks [{:size 40} {:size 40}]
    :64-bit true
    :io :high
    :ebs-optimized 500}

   :m3.2xlarge
   {:ram (GiB 30)
    :cpus [{:cores 8 :speed 3.25}]
    :disks [{:size 80} {:size 80}]
    :64-bit true
    :io :high
    :ebs-optimized 1000}

   ;;; Compute Optimized

   :c3.large
   {:ram (GiB 3.75)
    :cpus [{:cores 2 :speed 3.5}]
    :disks [{:size 16} {:size 16}]
    :64-bit true
    :io :moderate
    :ebs-optimized false}

   :c3.xlarge
   {:ram (GiB 7.5)
    :cpus [{:cores 4 :speed 3.5}]
    :disks [{:size 40} {:size 40}]
    :64-bit true
    :io :moderate
    :ebs-optimized 500}

   :c3.2xlarge
   {:ram (GiB 15)
    :cpus [{:cores 8 :speed 3.5}]
    :disks [{:size 80} {:size 80}]
    :64-bit true
    :io :high
    :ebs-optimized 1000}

   :c3.4xlarge
   {:ram (GiB 30)
    :cpus [{:cores 16 :speed 3.4375}]
    :disks [{:size 160} {:size 160}]
    :64-bit true
    :io :high
    :ebs-optimized 2000}

   :c3.8xlarge
   {:ram (GiB 60)
    :cpus [{:cores 32 :speed 3.375}]
    :disks [{:size 320} {:size 320}]
    :64-bit true
    :io :very-high
    :ebs-optimized false}

   ;;; GPU Instances

   :g2.2xlarge
   {:ram (GiB 15)
    :cpus [{:cores 8 :speed 3.25}]
    :disks [{:size 60}]
    :64-bit true
    :io :high
    :ebs-optimized 1000}

   ;;; Memory Optimized

   :r3.large
   {:ram (GiB 15)
    :cpus [{:cores 2 :speed 3.25}]
    :disks [{:size 32}]
    :64-bit true
    :io :moderate
    :ebs-optimized false}

   :r3.xlarge
   {:ram (GiB 30.5)
    :cpus [{:cores 4 :speed 3.25}]
    :disks [{:size 80}]
    :64-bit true
    :io :moderate
    :ebs-optimized 500}

   :r3.2xlarge
   {:ram (GiB 61)
    :cpus [{:cores 8 :speed 3.25}]
    :disks [{:size 160}]
    :64-bit true
    :io :high
    :ebs-optimized 1000}

   :r3.4xlarge
   {:ram (GiB 122)
    :cpus [{:cores 16 :speed 3.25}]
    :disks [{:size 320}]
    :64-bit true
    :io :high
    :ebs-optimized 2000}

   :r3.8xlarge
   {:ram (GiB 244)
    :cpus [{:cores 32 :speed 3.25}]
    :disks [{:size 320} {:size 320}]
    :64-bit true
    :io :very-high
    :ebs-optimized false}

   ;;; Storage Optimized

   :i2.xlarge
   {:ram (GiB 30.5)
    :cpus [{:cores 4 :speed 3.5}]
    :disks [{:size 800}]
    :64-bit true
    :io :moderate
    :ebs-optimized 500}

   :i2.2xlarge
   {:ram (GiB 61)
    :cpus [{:cores 8 :speed 3.375}]
    :disks [{:size 800} {:size 800}]
    :64-bit true
    :io :high
    :ebs-optimized 1000}

   :i2.4xlarge
   {:ram (GiB 122)
    :cpus [{:cores 16 :speed 3.3125}]
    :disks [(repeat 4 {:size 800})]
    :64-bit true
    :io :high
    :ebs-optimized 2000}

   :i2.8xlarge
   {:ram (GiB 244)
    :cpus [{:cores 32 :speed 3.25}]
    :disks [(repeat 8 {:size 800})]
    :64-bit true
    :io :very-high
    :ebs-optimized false}

   :hs1.8xlarge
   {:ram (GiB 117)
    :cpus [{:cores 16 :speed 2.1875}]
    :disks [(repeat 24 {:size 2048})]
    :64-bit true
    :io :very-high
    :ebs-optimized false}

   ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
   ;;; Previous generation instance types

   :m1.small
   {:ram (* 1024 1.7)                   ; GiB memory
    :cpus [{:cores 1 :speed 1}]         ; 1 EC2 Compute Unit
    :disks [{:size 160}]                ; GB instance storage
    :32-bit true
    :64-bit true
    :io :moderate
    :ebs-optimised false}

   :m1.medium
   {:ram (* 1024 3.75)                  ;  GiB memory
    :cpus [{:cores 1 :speed 2}]         ; EC2 Compute Unit
    :disks [{:size 410}]                ; GB instance storage
    :32-bit true
    :64-bit true
    :io :moderate
    :ebs-optimised false}

   :m1.large
   {:ram (* 1024 7.5)                   ; GiB memory
    :cpus [{:cores 2 :speed 2}]         ; 4 Compute Units
    :disks [{:size 850}]                ; GB instance storage
    :64-bit true
    :io :high
    :ebs-optimised 500}                 ; Mbps

   :m1.xlarge
   {:ram (* 1024 15)                    ; GiB memory
    :cpus [{:cores 4 :speed 2}]         ; 8 Compute Units
    :disks [{:size 1690}]               ; GB instance storage
    :64-bit true
    :io :high
    :ebs-optimised 1000}                ; Mbps

   :t1.micro
   {:ram 613                            ; MiB memory
    :cpus [{:cores 1 :speed 1}]         ; Up to 2 EC2 Compute Units
    :disks []
    :32-bit true
    :64-bit true
    :io :low
    :ebs-optimised false}

   :m2.xlarge
   {:ram (* 1024 17.1)                  ; GiB of memory
    :cpus [{:cores 2 :speed 3.25}]      ; 6.5 Compute Units
    :disks [{:size 420}]                ; GB of instance storage
    :64-bit true
    :io :moderate
    :ebs-optimised false}

   :m2.2xlarge
   {:ram (* 1024 34.2)                  ; GiB of memory
    :cpus [{:cores 4 :speed 3.25}]      ; 13 Compute Units
    :disks [{:size 850}]                ; GB of instance storage
    :64-bit true
    :io :high
    :ebs-optimised false}

   :m2.4xlarge
   {:ram (* 1024 68.4)                  ; GiB of memory
    :cpus [{:cores 8 :speed 3.25}]      ; 26 Compute Units
    :disks [{:size 1690}]               ; GB of instance storage
    :64-bit true
    :io :high
    :ebs-optimised 1000}                ; Mbps

   :c1.medium
   {:ram (* 1024 1.7)                   ; GiB of memory
    :cpus [{:cores 2 :speed 2.5}]       ; 5 Compute Units
    :disks [{:size 350}]                ; GB of instance storage
    :32-bit true
    :64-bit true
    :io :moderate
    :ebs-optimised false}

   :c1.xlarge
   {:ram (* 1024 7)                     ; GiB of memory
    :cpus [{:cores 8 :speed 2.5}]       ; 20 EC2 Compute Units
    :disks [{:size 1690}]               ; GB of instance storage
    :64-bit true
    :io :high
    :ebs-optimised false}

   :cc1.4xlarge
   {:ram (* 1024 23)                               ; GiB of memory
    :cpus [{:cores 4 :speed 4}{:cores 4 :speed 4}] ; 33.5 EC2 Compute Units
    :disks [{:size 1690}]               ; GB of instance storage
    :64-bit true
    :io :very-high                      ; (10 Gigabit Ethernet)
    :ebs-optimised false}

   :cc2.8xlarge
   {:ram (* 1024 60.5)                             ; GiB of memory
    :cpus [{:cores 8 :speed 4}{:cores 8 :speed 4}] ; 88 EC2 Compute Units
    :disks [{:size 3370}]               ; GB of instance storage
    :64-bit true
    :io :very-high                      ; (10 Gigabit Ethernet)
    :ebs-optimised false}

   :cg1.4xlarge
   {:ram (* 1024 22)                               ; GiB of memory
    :cpus [{:cores 4 :speed 4}{:cores 4 :speed 4}] ; 33.5 EC2 Compute Units
    :disks [{:size 1690}]               ; GB of instance storage
    :64-bit true
    :io :very-high                      ; (10 Gigabit Ethernet)
    :ebs-optimised false}

   :hi1.4xlarge
   {:ram (* 1024 60.5)                  ; GiB of memory
    :cpus [{:cores 16 :speed 2.1875}]   ; 35 EC2 Compute Units
    :disks [{:size 1024}{:size 1024}] ; 2 SSD-based volumes each with 1024 GB
                                        ; of instance storage
    :64-bit true
    :io :very-high                      ; (10 Gigabit Ethernet)
    :ebs-optimised false}

   :hi1.8xlarge
   {:ram (* 1024 117)                   ; GiB of memory
    :cpus [{:cores 16 :speed 2.1875}]   ; 35 EC2 Compute Units
    :disks (repeat 24 {:size 1024}) ; 24 SSD-based volumes each with 1024 GB
                                        ; of instance storage
    :64-bit true
    :io :very-high                      ; (10 Gigabit Ethernet)
    :ebs-optimised false}
   })


src/pallet/compute/ec2/ami.clj
(ns pallet.compute.ec2.ami
  "Functions for parsing ami data"
  (:require
   [clojure.core.match :refer [match]]
   [clojure.core.match.regex]
   [clojure.string :as string]))

;;; ### OS parsing
(def ^{:doc "Map of non-canonical os-versions to canonical versions"}
  os-versions
  {"debian" {"lenny" "5.0" "squeeze" "6.0" "5" "5.0" "6" "6.0"}
   "centos" {"5" "5.0" "6" "6.0"}
   "rhel" {"5" "5.0" "6" "6.0"}
   "ubuntu" {"hardy" "8.04", "karmic" "9.10", "lucid" "10.04",
             "maverick" "10.10", "natty" "11.04", "oneiric" "11.10",
             "precise" "12.04"}
   "windows" {"2008 Server R2" "2008 R2",
              "Server 2008" "2008",
              "Server 2008 SP2" "2008 SP2",
              "2008 Web" "2008",
              "2008 R2" "2008 R2",
              "2008 SP2" "2008 SP2",
              "2003 Standard" "2003",
              "Server 2008 R2" "2008 R2",
              "2008 Server" "2008"}})

(def ^{:doc "Map of non-canonical os-families to canonical versions"}
  os-families
  {})

;; this is based on jclouds' AWSEC2ReviseParsedImage
(defn default-parser-result
  [match]
  (when-let [[_ family version] match]
    (let [family (string/lower-case family)
          versions (os-versions family {})]
      {:os-family (keyword (os-families family family))
       :os-version (versions version version)})))

(def os-parsers
  {:amzn-linux
   #".*amzn-(?:hvm-)?ami-(?:pv-)?(.*)\.(i386|x86_64)(-ebs|\.manifest.xml)?"
   :amazon #"amazon/EC2 ([^ ]+) ([^ ]+).*"
   :canonical #".*/([^-]*)-([^-]*)-.*-(.*)(\\.manifest.xml)?"
   :rightimage
   #"[^/]*/RightImage[_ ]([^_]*)_([^_]*)_[^vV]*[vV](.*)(\.manifest.xml)?"
   :rightscale #"[^/]*/([^_]*)_([^_]*)_[^vV]*[vV](.*)(\.manifest.xml)?"})

(defn parse
  "Best guess os family and version from the image name"
  [{:keys [name] :as image}]
  (match [image]
    [{:owner-id "137112412989"
      :description #"Amazon Linux AMI.*"}]
    (let [[m version] (re-find (:amzn-linux os-parsers) name)]
      {:os-family :amzn-linux
       :os-version version
       :user {:username "ec2-user"}})

    [{:owner-id "137112412989"}]
    (default-parser-result (re-find (:amazon os-parsers) name))

    [{:owner-id "099720109477"}]
    (assoc (default-parser-result (re-find (:canonical os-parsers) name))
      :user {:username "ubuntu"})

    [{:owner-id "411009282317"
      :name #".*RightImage.*"}]
    (default-parser-result (re-find (:rightimage os-parsers) name))

    [{:owner-id "411009282317"}]
    (default-parser-result (re-find (:rightscale os-parsers) name))))

;; local Variables:
;; mode: clojure
;; eval: (define-clojure-indent (match 1))
;; End:
